sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SerializationConstructorsShouldBeSecured.cs

```diff
@@ -0,0 +1,139 @@
+
+        private static readonly AttributeComparer attributeComparer = new AttributeComparer();
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(FindPossibleViolations,
+                SyntaxKind.ConstructorDeclaration);
+        }
+
+        private void FindPossibleViolations(SyntaxNodeAnalysisContext c)
+        {
+            var constructorSyntax = (ConstructorDeclarationSyntax)c.Node;
+            var reportLocation = constructorSyntax?.Identifier.GetLocation();
+            if (reportLocation == null)
+            {
+                return;
+            }
+
+            var serializationConstructor = c.SemanticModel.GetDeclaredSymbol(constructorSyntax);
+            if (!serializationConstructor.IsSerializationConstructor())
+            {
+                return;
+            }
+
+            var classSymbol = serializationConstructor.ContainingType;
+            if (!classSymbol.Implements(KnownType.System_Runtime_Serialization_ISerializable))
+            {
+                return;
+            }
+
+            var isAssemblyIsPartiallyTrusted = c.SemanticModel.Compilation.Assembly
+                    .GetAttributes()
+                    .Any(a => a.AttributeClass.Is(KnownType.System_Security_AllowPartiallyTrustedCallersAttribute));
+            if (!isAssemblyIsPartiallyTrusted)
+            {
+                return;
+            }
+
+            var serializationConstructorAttributes = GetCasAttributes(serializationConstructor);
+
+            bool isConstructorMissingAttributes = classSymbol.Constructors
+                .SelectMany(m => GetCasAttributes(m))
+                .Any(attr => !serializationConstructorAttributes.Contains(attr, attributeComparer));
+
+            if (isConstructorMissingAttributes)
+            {
+                c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, reportLocation));
+            }
+        }
+
+        private static IEnumerable<AttributeData> GetCasAttributes(IMethodSymbol methodSymbol)
+        {
+            return methodSymbol.GetAttributes().Where(IsCasAttribute);
+
+            bool IsCasAttribute(AttributeData data) =>
+                data?.AttributeClass.DerivesFrom(KnownType.System_Security_Permissions_CodeAccessSecurityAttribute)
+                    ?? false;
+        }
+
+        private class AttributeComparer : IEqualityComparer<AttributeData>
+        {
+            public bool Equals(AttributeData x, AttributeData y)
+            {
+                return Equals(x.AttributeConstructor, y.AttributeConstructor) &&
+                    Enumerable.SequenceEqual(x.ConstructorArguments, y.ConstructorArguments) &&
+                    AreNamedArgumentsEqual(x.NamedArguments, y.NamedArguments);
+            }
+
+            private bool AreNamedArgumentsEqual(
+                ImmutableArray<KeyValuePair<string, TypedConstant>> argumentsX,
```

Comment:
This doesn't need to be `ImmutableArray`, please use a less derived type.

Reply:
Good point. S3242 should have pointed it out, I'll check why it did not.

Reply:
This is actually fine. Extensions on `ImmutableArray` are in `ImmutableArrayExtensions`, not `Enumerable` as I'd expect.